deepinv/loss/score.py

```diff
@@ -1,41 +1,161 @@
+
+    :param None, torch.nn.Module noise_model: Noise distribution corrupting the measurements
+        (see :ref:`the physics docs <physics>`). Options are :class:`deepinv.physics.GaussianNoise`,
+        :class:`deepinv.physics.PoissonNoise`, :class:`deepinv.physics.GammaNoise` and
+        :class:`deepinv.physics.UniformGaussianNoise`. By default, it uses the noise model associated with
+        the physics operator provided in the forward method.
+    :param int total_batches: Total number of training batches (epochs * number of batches per epoch).
+    :param tuple delta: Tuple of two floats representing the minimum and maximum noise level,
+        which are annealed during training.
+
+    |sep|
+
+
+    :Example:
+
+        >>> import torch
+        >>> import deepinv as dinv
+        >>> sigma = 0.1
+        >>> physics = dinv.physics.Denoising(dinv.physics.GaussianNoise(sigma))
+        >>> model = dinv.models.DnCNN(depth=2, pretrained=None)
+        >>> loss = dinv.loss.ScoreLoss(total_batches=1, delta=(0.001, 0.1))
+        >>> model = loss.adapt_model(model) # important step!
+        >>> x = torch.ones((1, 3, 5, 5))
+        >>> y = physics(x)
+        >>> x_net = model(y, physics, update_parameters=True) # save score loss in forward
+        >>> l = loss(model)
+        >>> print(l.item() > 0)
+        True
     """
 
-    def __init__(self, delta):
+    def __init__(self, noise_model=None, total_batches=1000, delta=(0.001, 0.1)):
         super(ScoreLoss, self).__init__()
-        self.name = "score"
-        self.metric = torch.nn.MSELoss()
+        self.total_batches = total_batches
         self.delta = delta
+        self.noise_model = noise_model
 
-    def forward(self, y, model, **kwargs):
+    def forward(self, model, **kwargs):
         r"""
-        Computes the Score loss.
+        Computes the Score Loss.
 
-        :param torch.Tensor y: measurements.
-        :param torch.nn.Module model: Reconstruction function.
-        :return: (torch.Tensor) loss.
+        :param torch.Tensor y: Measurements.
+        :param deepinv.physics.Physics physics: Forward operator associated with the measurements.
+        :param torch.nn.Module model: Reconstruction model.
+        :return: (torch.Tensor) Score loss.
         """
-        std = np.randn() * self.delta
-        noise = torch.randn_like(y)
-        return self.metric(noise, std * model(std * noise + y))
+        return model.get_error()
+
+    def adapt_model(self, model, **kwargs):
+        r"""
+        Transforms score backbone net :meth:`S` into :meth:`R` for training and evaluation.
+
+        :param torch.nn.Module model: Backbone model approximating the score.
+        :return: (torch.nn.Module) Adapted reconstruction model.
+        """
+        if isinstance(model, ScoreModel):
+            return model
+        else:
+            return ScoreModel(model, self.noise_model, self.delta, self.total_batches)
+
+
+class ScoreModel(torch.nn.Module):
```

Comment:
Missing some docstring for the arguments

Reply:
now it should be fine right?